### DISCUSSION TITLE
Help with relationship field type - inline_create array

### DISCUSSION THREAD

**Discussion by MaraIGI** [Laravel-Backpack/community-forum#717] [Q&A (Help)]
Hello there,

I'm in the process of implementing Backpack CRUD for client's custom needs, and I've come to some difficulties trying to implement InlineCreateOperation (https://backpackforlaravel.com/docs/6.x/crud-operation-inline-create)

I have these scheme of Models:
- Instance
  - InstanceVersion
    - InstanceVersionDebt

As you can see, I tried for you to show, that nested logic which I work with, I have multiple Instances, with HasMany InstanceVersions, and every InstanceVersion HasMany InstanceVersionDebts.

I want `select2 multiple field`, to select InstanceVersionDebts, which I will create in Modal, and have that `+ Add` link on top right position of my select2 field. Currently it drives me crazy, because for some hours which I put in I still can't achieve that - that little `+ Add` thing.

I would be really glad if someone could help me out - I tried to post below all codes, which I considered important parts connected with the problem.

I think problem is in my Model names, CRUD classes names or routes - which comes out from Model names above. Also 'instance-verze-dluhy' below is Czech pretty-slug name for that Model, CRUD route.

My `InstanceVersionDebtCrudController` looks like this:
```
..
use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation;
use \Backpack\Pro\Http\Controllers\Operations\InlineCreateOperation;
..
public function setup() {
  CRUD::setModel(InstanceVersionPerson::class);
  CRUD::setRoute(config('backpack.base.route_prefix') . '/instance-verze-dluhy');
  CRUD::setEntityNameStrings('instance - verze - dluh', 'instance - verze - dluhy');
}
..
protected function setupCreateOperation() {
  CRUD::setFromDb(); // like in blank project, I just need concept with which I can later work
}
```

And my `InstanceVersionCrudController` looks like this:
```
..
use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation;
use \Backpack\Pro\Http\Controllers\Operations\FetchOperation;
..
protected function fetchInstanceVersionDebts() {
  return $this->fetch(InstanceVersionDebt::class);
}
..
public function setupCreateOperation() {
CRUD::addField([
  'name' => 'instanceVersionDebts',
  'label' => 'Clients debts',
  'type' => 'relationship',
  'ajax' => true,
  'inline_create' => [
    'entity' => 'instanceVersionDebt', // I absolutly don't know what to set here
    'force_select' => true,
    'modal_class' => 'modal-dialog modal-xl',
    // I tried to put manually the route, it works, InlineOperation creates it
    'modal_route' => route('instance-verze-dluhy-inline-create'),
    'create_route' =>  route('instance-verze-dluhy-inline-create-save'),
    'add_button_label' => 'Add',
  ],
]);
}
```

And my Models, I hope, are configured right: ..

Model `InstanceVersion`:
```
// has own `id` column, attribute
..
public function instanceVersionDebts(): HasMany {
  return $this->hasMany(InstanceVersionDebt::class);
}
..
```

Model `InstanceVersionDebt`:
```
// also has `id` column, attribute, but it's not important here
// mainly has `instance_version_id` column, atribute, which is constrained via Eloquent, Laravel migration files
// .. to the InstanceVersion Model
..
public function instanceVersion(): BelongsTo {
  return $this->belongsTo(InstanceVersion::class);
}
..
```

I looked at docs (https://backpackforlaravel.com/docs/6.x/crud-fields) and also at Github codes of demo (https://github.com/Laravel-Backpack/demo/blob/main/app/Http/Controllers/Admin/MonsterCrudController.php) but still couldn't figured it out.

Will be glad for any help, thanks in advance,

Mark

**Comment by pxpm** [comment]
That `InlineCreate`  operation needs some love, thanks for the report @MaraIGI 

`entity` is "misleading" in that operation because it's confusing with the "entity" in fields/columns, that represent the relation string.

Not the case here. "entity" there means the "route entity" that we are going to use. 

Imagine you have a `tags` relationship:
```php
CRUD::field('tags')->inline_create(['entity' => 'tag']);
```
This means that you want the `inline_create`, to get the Create Form from `site.com/admin/**tag**/create` instead of `tags`.

Can you post a screenshot of the error your are getting ? I am guessing is some messup with entity names/routes. 

Apart from the general [Operation Docs](https://backpackforlaravel.com/docs/6.x/crud-operation-inline-create#about-1), there is also [this small snippet](https://backpackforlaravel.com/docs/6.x/crud-fields#create-related-entries-in-a-modal-using-the-inlinecreate-operati) with useful information.

Cheers

**Reply by MaraIGI** [reply]
Hi Pedro! Thanks for your comment, and sorry for my answer-delay, the docs which you sent I already went up and down, with some tries and unfortunate failures.

Also in some point I figured out that the `entity` attribute for adding fields is misleading, and would meant that its for that slugged url part `../**tag**/create` and etc., like in my case, with my example, `InstanceVersionDebt` Model, where the slugged part is `instance-verze-dluhy` (translated to pretty-slug to the Czech language)

The problem is, the `inline_create` operation, Fields attribute, with `relationship` type, doesn't give me any errors, at all, and doesn't show me that little sweet link with "+ Add" (in my case as could be seen below in code, customized, "Create new InstanceVersionDebt") which I'm trying to achieve.

And yes, I'm clearing cache responsibly - when I add any operations to CRUD Controller and when doing any changes to custom.php file, so Laravel router can setup it's routes right.

I, once more time, summarized everything to one place so it's even more clear, and added the right entity to it, and I'm attaching screenshot from it's render:

Screenshot from `InstanceVersionCrudController`, with missing `+ Add` which should be above `x Vyčistit` string
![Screenshot from InstanceVersionCrud](https://img001.prntscr.com/file/img001/E9rZzEf_Qb2OEFNbcLtvGw.png)


It's on my hastebin here: [https://hastebin.vajsbejn.cz/nevuyacusa.php](https://hastebin.vajsbejn.cz/nevuyacusa.php), and after thinking I'm also pasting it here again in one code block: 

```

Save New Duplicate & Edit Just Text
// InstanceVersion.php
class InstanceVersion extends Model
{
    use CrudTrait;
    ..
    ..
    // this fully works, I can even do 'relationship' on it with fully working subfields
    public function instanceVersionDebts(): HasMany
    {
        return $this->hasMany(InstanceVersionDebt::class);
    }
    ..
}

// InstanceVersionDebt.php
class InstanceVersionDebt extends Model
{
    use CrudTrait;
    ..
    ..
    // this Model has only this relation
    public function instanceVersion(): BelongsTo
    {
        return $this->belongsTo(InstanceVersion::class);
    }
}

// InstanceVersionCrudController.php
class InstanceVersionCrudController extends CrudController
{
    ..
    ..
    use \Backpack\Pro\Http\Controllers\Operations\FetchOperation;
    ..
    protected function fetchInstanceVersionDebts()
    {
        return $this->fetch(InstanceVersionDebt::class);
    }
    ..
    ..
    protected function setupCreateOperation()
    {
	CRUD::addField([
	    'tab'           => 'InstanceVersionDebt Model',
	    'name'          => 'instanceVersionDebts',
	    'label'         => 'instanceVersionDebts relation',
	    'type'          => 'relationship',
	    'ajax'          => true,
	    'inline_create' => [
		'entity'        => 'instance-verze-dluhy',
		'force_select' => true,
		'modal_class' => 'modal-dialog modal-xl',
		'modal_route' => route('instance-verze-dluhy-inline-create'),
		'create_route' =>  route('instance-verze-dluhy-inline-create-save'),
	        'add_button_label' => 'Create new InstanceVersionDebt',
	    ],
        ]);
    }
    ..
    ..
}

// InstanceVersionDebtCrudController.php
class InstanceVersionDebtCrudController extends CrudController
{
    use \Backpack\CRUD\app\Http\Controllers\Operations\ListOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\UpdateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\DeleteOperation;
    //use \Backpack\CRUD\app\Http\Controllers\Operations\ShowOperation;
    ..
    use \Backpack\CRUD\app\Http\Controllers\Operations\InlineCreateOperation;
    ..
    ..
    public function setup()
    {
        CRUD::setModel(InstanceVersionDebt::class);
        CRUD::setRoute(config('backpack.base.route_prefix') . '/instance-verze-dluhy');
        CRUD::setEntityNameStrings('instanci - verzi - dluh', 'instance - verze - dluhy');
    }
    ..
    ..
    protected function setupCreateOperation()
    {
      // shrinked it to only one field
        CRUD::addField([
	    'name'      => 'name',
	    'label'     => 'Debt name',
	    'type'      => 'text',
	    'wrapper'   => [ 'class' => 'form-group', ],
	]);
    }
	
    protected function setupInlineCreateOperation()
    {
        $this->setupCreateOperation();
    }
    ..
}
	
// custom.php routes
..
Route::crud('instance-verze', 'InstanceVersionCrudController');
Route::crud('instance-verze-dluhy', 'InstanceVersionDebtCrudController');
..
```

At this moment I really don't know if I'm missing anything or it could be any technical problem from Backpack, thanks for yours any further help, I'm looking to finishing this feature up and my clients will be very happy if it will work at the end 🙏

Also it's my only working solution in my use-case. I'm actually not trying to implement it on this `InstanceVersionDebt` model, but on Models, situation, which has this hierarchy, relations:

`InstanceVersion HasMany InstanceIncome and InstanceVersionIncome HasMany InstanceVersionPerson (and InstanceVersionPerson has to be also linked to InstanceVersion to overcome some problems)` .. but until I figure it out on simple example, I can't move to this one - which will at the end be very similar to this Debts example.

**Reply by pxpm** [reply]
Hey @MaraIGI thanks for the response back. 🙏  
I think the issue is that `InlineCreate` does not work on `HasMany` relations. 
Theoretically it could work if we re-factored the way `InlineCreate` works, but at the moment it's only possible to use `InlineCreate` in `BelongsTo` and `BelongsToMany` relations. 

To create multiple instances of the same relation on the form (for HasMany relations), you can use the `subfields` of the relationship: https://backpackforlaravel.com/docs/6.x/crud-how-to#hasmany-1-n-relationship-with-subform-to-create-update-and-delet-1

Let me know if that was the issue. 

Cheers

**Reply by MaraIGI** [reply]
Ah, I see, read more read more read more, .. it's in the documentation of [Fields, right here](https://backpackforlaravel.com/docs/6.x/crud-fields#create-related-entries-in-a-modal-using-the-inlinecreate-operati), but not in documentation of [InlineCreateOperation](https://backpackforlaravel.com/docs/6.x/crud-operation-inline-create). But that's again something, that deepened more knowledge about Backpack and it's usage, and about to what I should focus in the docs.

To zoom in on my use-case which I want to achieve, as I said, I have:
`InstanceVersion HasMany InstanceIncome and InstanceVersionIncome HasMany InstanceVersionPerson (and InstanceVersionPerson has to be also linked to InstanceVersion to overcome some problems)`

I wanted to have my client option to add `People` via the `+ Add` option and in no time add that Person to database (yes - it's achievable only when `InstanceVersion` is created, so these tabs - `Incomes` and `People - editation` are visible in update form, if I won't fix this issue I'm maybe adding `People - editation` tab, it's content, to createOperation too). And I wanted to be the Person searchable right in the select2 in other field on the same page without page refresh needed (ajax).

_That's the reason why I can't use subfields here, I think. I would need something to push new records which I create via subfields method in no time to database._

Also I bumped to issue, when client couldn't edit the added `Persons` detail in same form, so I created a new Tab `People - editation` in edit form, but it's edits aren't saved until form save, and when client would add client there (which could be permitted, now its not on screenshots below), it won't be visible in that select on `Incomes` tab.

I hope I didn't described it too much confusingly - maybe it could be some feature which could be useful and added to the core :-)

Thanks for your reply Pedro 🙏 and clarification about this `InlineCreateOperation` which I overlooked.

Also, to solve my problem, I don't know exactly what to do, I will look to `BelongsToMany` relation which doesn't give me too much sense (maybe I didn't get it's purpose right, right now).

![Screen 1 from admin](https://img001.prntscr.com/file/img001/nzmhDPJrSDSIvo28bwej6g.png)

![Screen 2 from admin](https://img001.prntscr.com/file/img001/g_9HAQAwT1uZVWG7lzIw5A.png)
